### PR TITLE
Add support for recursive Coders

### LINKED ISSUE #1992: Coder derivation for nested case class
Here is my simplified case class, when I try to generate Coder for it, different behaviours are observed, can someone help explain them ? Thanks!
```scala
import com.spotify.scio.coders.Coder

case class SampleField(name: String, fieldType: SampleFieldType)

sealed trait SampleFieldType
case object IntegerType extends SampleFieldType
case object StringType extends SampleFieldType
case class RecordType(fields: Vector[SampleField]) extends SampleFieldType
```
**Compiler error when trying to derive coders**

```scala
implicit val fieldTypeCoder = Coder[SampleFieldType]  
implicit val fieldCoder= Coder[SampleField]
// Cannot find an implicit Coder instance for type:
// [error] 
// [error]   >> pipeline.SampleField
//...
```
**Coders are null when type annotations are added**

```scala
implicit val fieldTypeCoder: Coder[SampleFieldType] = Coder[SampleFieldType]  // fieldTypeCoder: null
implicit val fieldCoder: Coder[SampleField] = Coder[SampleField] // fieldCoder: null
```

**Coders derived successfully when a random class added in between**
```scala
implicit val fieldTypeCoder = Coder[SampleFieldType]  // fieldTypeCoder: Disjunction(pipeline.SampleFieldType, Map(0 -> Record(pipeline.IntegerType, ), 1 -> Record(pipeline.RecordType, (fields, Fallback(scala.collection.immutable.Vector))), 2 -> Record(pipeline.StringType, )))

case class Foo()  
implicit val fooCoder = Coder[Foo]  // fooCoder: Record(pipeline.Pipeline.Foo, )
  
implicit val fieldCoder= Coder[SampleField] // fieldCoder: Record(pipeline.SampleField, (name, Beam(StringUtf8Coder)), (fieldType, Disjunction(pipeline.SampleFieldType, Map(0 -> Record(pipeline.IntegerType, ), 1 -> Record(pipeline.RecordType, (fields, Fallback(scala.collection.immutable.Vector))), 2 -> Record(pipeline.StringType, )))))
```

### PR DESCRIPTION
**Pull request by jto**
fixes #1992

### CODE REVIEW

File: scio-test/src/test/scala/com/spotify/scio/coders/coders.scala [hunk 1]
```diff
@@ -455,4 +455,17 @@ class CodersTest extends FlatSpec with Matchers {
     val wrappedArray: mut.WrappedArray[String] = Array("foo", "bar")
     wrappedArray coderShould notFallback()
   }
+
+  it should "support derivation of recursive types" in {
+    case class SampleField(name: String, fieldType: SampleFieldType)
+    sealed trait SampleFieldType
+    case object IntegerType extends SampleFieldType
+    case object StringType extends SampleFieldType
+    case class RecordType(fields: Vector[SampleField]) extends SampleFieldType
```
Comment (regadas) [inline scio-test/src/test/scala/com/spotify/scio/coders/coders.scala]: in this test replacing `Vector` with `List` and getting as well a Coder[RecordType] should tigger the issue

Comment (jto) [reply scio-test/src/test/scala/com/spotify/scio/coders/coders.scala]: It does. I'm working on that example and trying to come up with a fix.

File: scio-macros/src/main/scala/com/spotify/scio/MagnoliaMacros.scala [hunk 2]
```diff
@@ -34,13 +34,7 @@ private[scio] object MagnoliaMacros {
...
     }
 
-    val magTree = magnolia.Magnolia.gen[T](c)
-
-    def getLazyVal =
-      magTree match {
-        case q"val $name = $body; $rest" =>
-          body
-      }
+    val getLazyVal = magnolia.Magnolia.gen[T](c)
```
Comment (regadas) [inline scio-macros/src/main/scala/com/spotify/scio/MagnoliaMacros.scala]: 🎨 should we rename this ... or even move it to `removeAnnotations.transform( ... )`?

File: scio-core/src/main/scala/com/spotify/scio/coders/Coder.scala [hunk 3]
```diff
@@ -60,6 +60,18 @@ Cannot find an implicit Coder instance for type:
 """
 )
 sealed trait Coder[T] extends Serializable
+
+private[scio] final class Ref[T](val typeName: String, c: => Coder[T]) extends Coder[T] {
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/coders/Coder.scala]: 🎨 replace `c` with `coder`?

File: scio-core/src/main/scala/com/spotify/scio/coders/Coder.scala [hunk 4]
```diff
@@ -60,6 +60,18 @@ Cannot find an implicit Coder instance for type:
...
+  def value = c
+
+  override def toString(): String = s"""Ref($typeName)"""
+}
+
+private[scio] object Ref {
+  def apply[T](t: String, c: => Coder[T]) = new Ref[T](t, c)
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/coders/Coder.scala]: 🎨 better arg names here as well?

**Pull request merged**